### ISSUE TITLE
Problem 1.1

### ISSUE THREAD

**Issue by ppaleja** [ppaleja/llmsys_f25_hw5#15]
- [ ] 1\. Implement the helper function `partition_dataset` in
`data_parallel/dataset.py`:

``` {.Python bgcolor="LightGray"}
def partition_dataset(
    dataset, batch_size=128, collate_fn=None
) -> Tuple[DataLoader, int]:
    ...
```

Hint:

1.  Calculate the partitioned batch size. We define
    `partitioned batch size` as the batch size on every individual
    device. The `partitioned batch size` for each GPU, for instance,
    would be $128 // 4 = 32$ if we have four GPUs and the total batch
    size is 128.

2.  Create a partitioner class `DataPartitioner` with dataset and a list
    of partitioned sizes. The dataset's fraction that will be
    distributed among the parallel devices is represented by the list of
    partitioned sizes, which is a list of float values. For instance, if
    the dataset is divided equally over 4 GPUs, the list of `sizes`
    should look like this: $[0.25, 0.25, 0.25, 0.25]$. (`world_size`: This variable represents the total number of GPUs (or parallel processes) used for training.
    The `DataPartitioner` class has already been provided in the starter code. You need to create an instance of DataPartitioner by passing your dataset and a list of partition sizes.) 

4.  Get the current partition dataset given `rank` by using the `use`
    function in DataPartitioner. (`rank`: Each GPU (or process) is assigned a unique identifier called rank (e.g., 0, 1, 2, 3 in a 4-GPU setup).
    The rank determines which slice of the dataset the GPU will work on.)

5.  Wrap the dataset with `torch.utils.data.DataLoader`, remember to
    customize the `collate_fn` from our customized function defined in
    `project/utils.py`.

- [ ] 2. Implement dataset class `class Partition()` in
`data_parallel/dataset.py`:

``` {.Python bgcolor="LightGray"}
class Partition():
    def __init__(self, data, index):
        ...
    
    def __len__(self):
        ...
    
    def __getitem__(self, index):
        ...
```

and partition class `class DataPartitioner()` in
`data_parallel/dataset.py`:

``` {.Python bgcolor="LightGray"}
class DataPartitioner():
    def __init__(self, data, sizes=[0.7, 0.2, 0.1], seed=1234):
        ...

    def use(self, partition):
        ...
```

Hint:

1.  Create a list of indices for every data point and use `rng` to
    shuffle the indices. The indices should be integers. A simple way
    can be a list like this $[0, 1, ..., len(data)-1]$.

2.  Create different partitions of indices according to `sizes` and
    store in `self.partitions`. Let's suppose that we have 8 data
    points, and 2 partitions for the data. One way of `self.partitions`
    can be $[[0, 3, 6, 7], [1, 2, 4, 5]]$.

To summarize, the `partition_dataset` function is to help create a
training dataset partition you need for a single device within a cluster
of devices. The `Partition` class is used to define a dataset class to
return the data according to partitioned indices. The `DataPartitioner`
class is used to partition any datasets according to different workload
defined as `sizes`.

- [ ] Pass the test

``` {.Bash bgcolor="LightGray"}
python -m pytest -l -v -k "a5_1_1"
```

We simply test whether you distribute the data into different partitions
without overlapping.
